oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelineMongoDownloadTask.java

```diff
@@ -0,0 +1,278 @@
+    private final int batchSize;
+    private final ArrayBlockingQueue<BasicDBObject[]> mongoDocQueue;
+    private final int retryDuringSeconds;
+    private final Logger traversalLog = LoggerFactory.getLogger(PipelineMongoDownloadTask.class.getName() + ".traversal");
+    private final MongoCollection<BasicDBObject> dbCollection;
+    private final ReadPreference readPreference;
+
+    private long documentsRead = 0;
+    private long nextLastModified;
+    private String lastIdDownloaded = null;
+
+    public <T extends Document> PipelineMongoDownloadTask(MongoDocumentStore mongoStore,
+                                                          Collection<T> collection,
+                                                          Predicate<String> filter,
+                                                          int batchSize,
+                                                          ArrayBlockingQueue<BasicDBObject[]> queue) {
+        this.mongoStore = mongoStore;
+        this.collection = collection;
+        this.batchSize = batchSize;
+        // TODO
+//        this.filter = filter;
+        this.mongoDocQueue = queue;
+
+        IndexingProgressReporter progressReporterPerTask =
+                new IndexingProgressReporter(IndexUpdateCallback.NOOP, NodeTraversalCallback.NOOP);
+        progressReporterPerTask.setMessagePrefix("Dumping from " + TRAVERSER_ID_PREFIX);
+
+        this.filter = (id) -> {
+            try {
+                progressReporterPerTask.traversedNode(() -> id);
+            } catch (CommitFailedException e) {
+                throw new RuntimeException(e);
+            }
+            traversalLog.trace(id);
+            return true;
+        };
+
+        // Default retries for 5 minutes.
+        this.retryDuringSeconds = ConfigHelper.getSystemPropertyAsInt(
+                OAK_INDEXER_PIPELINED_MONGO_CONNECTION_RETRY_SECONDS,
+                DEFAULT_OAK_INDEXER_PIPELINED_MONGO_CONNECTION_RETRY_SECONDS);
+        this.dbCollection = MongoDocumentStoreHelper.getDBCollection(mongoStore, collection);
+
+        //TODO This may lead to reads being routed to secondary depending on MongoURI
+        //So caller must ensure that its safe to read from secondary
+//        this.readPreference = MongoDocumentStoreHelper.getConfiguredReadPreference(mongoStore, collection);
+        this.readPreference = ReadPreference.secondaryPreferred();
+        LOG.info("Using read preference {}", readPreference.getName());
+
+    }
+
+    @Override
+    public Result call() throws Exception {
+        String originalName = Thread.currentThread().getName();
+        Thread.currentThread().setName("mongo-dump");
+        try {
+            LOG.info("Starting to download from MongoDB.");
+            MongoCollection<BasicDBObject> dbCollection = MongoDocumentStoreHelper.getDBCollection(mongoStore, collection);
+            //TODO This may lead to reads being routed to secondary depending on MongoURI
```

Comment:
I'm not sure. The MongoDocumentStore implementation has logic to limit the reads from secondaries if they are too far back. Maybe we can implement a similar logic here. But I don't know what could be the problems, maybe it will miss some of the recent documents?